README.md
# ddSeeker

## Description
**ddSeeker** extracts cellular and molecular identifiers from single cell RNA sequencing experiments.

**Input**: R1 and R2 FASTQ files from a paired-end single cell sequencing experiment.

**Output**: one unmapped BAM ([uBAM](https://gatkforums.broadinstitute.org/gatk/discussion/11008/ubam-unmapped-bam-format))
file containing reads tagged with cell barcodes and Unique Molecular Identifiers (UMI).
Default [tags](https://genome.sph.umich.edu/wiki/SAM#What_are_TAGs.3F) are
**XC** and **XM** for cell barcodes and UMI, **XE** for errors related to
the barcode identification, and XQ and Xq for base quality of cell barcode and UMI respectively[^1].
Users can manually set different tags (see [Additional options](https://github.com/cgplab/ddSeeker#additional-options)).

[^1]: Know issue. The cell barcode is composed by three shorter blocks and if
  any of the block is affected by and indel event it may be difficult to
  determine which base was inserted/deleted (although each block is fixed by
  comparing it to a list of available blocks). Therefore, the quality string of
  the cell barcode may differ in length with respect to the proper cell barcode
  by one base: it still gives a general view of the quality of the barcode.

### Errors in barcode identification

  - LX = both linkers not aligned correctly
  - L1 = linker 1 not aligned correctly
  - L2 = linker 2 not aligned correctly
  - I  = indel in BC2
  - D  = deletion in Phase Block or BC1
  - J  = indel in BC3 or ACG trinucleotide
  - K  = indel in UMI or GAC trinucleotide
  - B  = one BC with more than 1 mismatch

## Additional options

  - Increment number of CPU units (faster analysis) with `-c/--cores`.
  - Manually set tags with `--tag-bc`, `--tag-umi`, `--tag-bc-q`, `--tag-umi-q`, `--tag-error`.
  - Print uncompressed SAM file to standard output (allowing direct feeding to other tools for
    filtering, sorting etc.) with `-o/--output -` (note the `-` sign).
  - Generate two csv files reporting the number of reads per cell and the distribution
  of error tags specifying the path with `-s/--summary-prefix`.
  - Create plots from the csv summary files using `make_graphs.R` (see ).


## Installation & Usage
Clone the repository and add the folder to your PATH variable

    git clone https://github.com/cgplab/ddSeeker.git
    export PATH=<path_to_ddSeeker>:$PATH

### Dependencies
- [Python](https://www.python.org/downloads) (>= 3.5)
- [Biopython](http://biopython.org) (>= 1.71)
- [pysam](https://pysam.readthedocs.io) (>= 0.14)

We suggest to install python packages using [pip](https://pip.pypa.io/en/stable/installing/)
which should be already installed if you are using Python3 >= 3.4.

    pip install biopython
    pip install pysam

### Examples

- __ddSeeker with 20 cores__

      ddSeeker.py --input sampleA_R1.fastq.gz sampleA_R2.fastq.gz --output sampleA_tagged.bam --cores 20

- __Print to stdout and pipe to samtools for queryname sorting__

      ddSeeker.py -i sampleA_R* -c 20 -o - | samtools sort -no sampleA_tagged_qsorted.bam

### Test dataset
[https://github.com/cgplab/ddSeeker_example_dataset](https://github.com/cgplab/ddSeeker_example_dataset)

### Generate summary files and make graphs
Requires [R >=3.4](https://www.r-project.org/) and the [tidyverse](https://www.tidyverse.org/) package.
Three plots are generated: dot plot of error distribution, absolute count of reads per
cell, and cumulative distribution of reads per cell. The latter two report by default
the whole set of barcodes in the csv file. To limit the report to a lower
number, specify it from the command line.

    mkdir summary_folder
    ddSeeker.py -i sampleA_R* -c 20 -o sampleA_tagged.bam -s summary_folder/sampleA
    make_graphs.R summary_folder/sampleA 2000

### Integrating single cell analysis pipelines
Several pipelines have been developed to perform single cell analysis.
Below we describe the main steps required to integrate our tool with
[Drop-seq tools](http://mccarrolllab.com/dropseq/),
[scPipe](https://github.com/LuyiTian/scPipe) and
[dropEst](https://github.com/hms-dbmi/dropEst).

#### Drop-seq tools
Since Drop-seq tools was our choice for our analyses, we provide a ready-to-use bash
script.  Simply run

    ddSeeker_dropSeq_tools.sh [options] sampleA_R1.fastq.gz sampleA_R2.fastq.gz

to produce aligned tagged reads in BAM format. 
Table of Counts can be obtained using the `DigitalExpression` tool included in Drop-seq tools.

#### scPipe
scPipe requires one FASTQ file with cell barcodes and UMIs stored in the header
of each read record. To change the output of **ddSeeker** use the option
`--pipeline scpipe`.

    ddSeeker.py -i <read1.fastq.gz> <read2.fast.gz> -o <tagged_reads.fastq.gz -c 20 --pipeline scpipe

In addition, set `bc_len=18` and `UMI_len=8` with the `sc_exon_mapping()` function.

#### dropEst
dropEst can work with tagged BAM files. Simply make the BamTags match with the
ddSeeker tags specifying them in the config.xml file

    <BamTags>
        <cb>XC</cb>
        <umi>XM</umi>
    </BamTags>

## Citation
Romagnoli D\*, Boccalini G\*, Bonechi M, Biagioni C, Fassan P, Bertorelli R, De
Sanctis V, Di Leo A, Migliaccio I, Malorni L, Benelli M. *ddSeeker: a tool for
processing Bio-Rad ddSEQ single cell RNA-seq data*. BMC Genomics. 2018; 19:960

Link to the paper: [ddSeeker: a tool for processing Bio-Rad ddSEQ single cell RNA-seq data](https://rdcu.be/bekqj)


NEWS.md
VERSION 1.2.1
-------------------------
 - Move all scripts into `code` folder
 - Solve bug in `make_graphs.R`: (no limit on max cells)

VERSION 1.2.0
-------------------------
 - Added two additional tags that report base quality of Cell Barcode (XQ) and UMI (Xq)

VERSION 1.1.0
-------------------------
- Removed separate `barcode.txt` file which is now hard-coded into the script

Other stuff
-------------------------
- `split_bam.py`: a script to create separate bam files for every single cell.
  (requires [simplesam](https://github.com/mdshw5/simplesam))


test/test.sh
#!/bin/bash
set -eu

../ddSeeker.py unmapped.bam \
    -o tagged.bam \
    -s summary \
    -n 10


code/make_graphs.R
#!/usr/bin/Rscript
# 2020-03-25
# version 0.1
suppressMessages(library(tidyverse))

theme_set(theme_minimal())

argv <- commandArgs(trailingOnly=TRUE)

if (length(argv) == 1) {
    n_cells <- Inf
} else if (length(argv) == 2){
    n_cells <- as.integer(argv[2])
} else {
    stop(call. = FALSE,
         paste0("Wrong number of arguments.\n\n",
                "  Usage: make_graphs.R <summary-path> [max number of cells]\n\n"))
}

flags <- c("LX", "L1", "L2", "I", "D", "J", "K", "B", "PASS")
short_errors <- factor(flags, rev(flags))

fname1 <- paste0(argv[1], ".errors.csv")
errors <- read_tsv(fname1, col_types=cols())
errors <- mutate(errors, Error=factor(Error, rev(flags)))
p1 <- ggplot(errors, aes(Fraction, Error, color=Error, label=round(Fraction, 2))) +
    geom_point() +
    geom_segment(aes(x=0, xend=Fraction, y=Error, yend=Error)) +
    xlim(0, 1) +
    labs(title="Errors in barcode identification") +
    theme(axis.title.y=element_blank(),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_line(size=.25),
          panel.grid.minor.y = element_blank(),
          legend.position = "none")

fname2 <- paste0(argv[1], ".cell_barcodes.csv")
barcodes <- read_tsv(fname2, col_type = "cid") %>%
    mutate(Rank = row_number())

barcodes_df <- bind_rows(barcodes) %>%
  filter(Rank <= n_cells) %>%
  mutate(CumSum = cumsum(Count)) %>%
  mutate(CumFrac = CumSum/max(CumSum))

p2 <- ggplot(barcodes_df, aes(Rank, Count)) +
    geom_line() +
    labs(title="Absolute Read Count", x="Ranked Cells", y=expression("Read Count Per Cell")) +
    scale_x_continuous(labels=function(x) trimws(format(x, big.mark = ","))) +
    scale_y_continuous(trans="log10",
                       labels=function(x) trimws(format(x, big.mark = ","))) +
    theme(axis.text.y = element_text(angle=90, hjust=.5),
          panel.grid.major = element_line(size=.25),
          panel.grid.minor = element_blank(),
          legend.position = "none")

p3 <- ggplot(barcodes_df, aes(Rank, CumFrac)) +
    geom_line()+
    labs(title="Cumulative Fraction", x="Ranked Cells", y="Cumulative Fraction of Reads per Cell") +
    scale_x_continuous(labels=function(x) trimws(format(x, big.mark = ","))) +
    theme(panel.grid.major = element_line(size=.25),
          panel.grid.minor = element_blank(),
          legend.position = "none")

outfile <- paste0(argv[1], ".pdf")
pdf(outfile, width = 4*1.61803, height = 4)
print(p1)
print(p2)
print(p3)
graphics.off()
message("PDF produced: ", outfile)


code/split_bams.py
#!/usr/bin/env python3

# Date: 2019-03-22
# Author: Dario Romagnoli
# Script to split cell line bam files into single cell bamfiles

import simplesam
import time
from argparse import ArgumentParser
from os import mkdir
from os.path import exists
from os.path import join as join_path
from sys import argv, exit, stderr

def info(message):
    current_time = time.localtime()
    print('[{time}] {message}'.format(
        time=time.strftime('%Y-%m-%d %H:%M:%S', current_time),
        message=message), file=stderr)

def pretty(d):
    for key, value in d.items():
        print("* {key}: {value}".format(key=str(key), value=str(value)), file=stderr)

def main():
    parser = ArgumentParser(description = "Utility to split a tagged BAM/SAM " + \
            "file into separate SAM files for different barcodes.")
    parser.add_argument("input", help = "BAM/SAM tagged file")
    parser.add_argument("outdir", help = "Directory where to store the splitted SAM files")
    parser.add_argument("-t", "--tag-name", default="XC", help = "TAG name (default=XC)")
    parser.add_argument("-b", "--barcodes-file", help = "Selection of cell barcodes (default: all barcodes in BAM/SAM file)")
    args = parser.parse_args()

    mkdir(args.outdir)
    # user provided a subset of barcodes (create dictionary and set a flag)
    if args.barcodes_file:
        barcodes = dict.fromkeys(open(args.barcodes_file).read().split(), 0)
        subset_barcodes = True
    else:
        barcodes = dict()
        subset_barcodes = False

    in_samfile = open(args.input, "r")
    in_sam = simplesam.Reader(in_samfile)

    print("Selected TAG name: {name}".format(name=args.tag_name), file=stderr)
    info("Analyzing file {file}".format(file=args.input))

    selected_reads = 0
    for tot_reads, read in enumerate(in_sam):
        if (tot_reads + 1) % 1e5 == 0:
            info("Reads (selected/total): {s}/{t}".format(s=str(selected_reads+1), t=str(tot_reads+1)))

        cell_bc = read.tags.get(args.tag_name)
        if (cell_bc and subset_barcodes and (cell_bc in barcodes)) or (cell_bc and not subset_barcodes):
            selected_reads += 1
            out_filename = join_path(args.outdir, cell_bc + ".sam")
            if not barcodes.get(cell_bc): # NB: both get zero and get null equal false
                barcodes[cell_bc] = barcodes.get(cell_bc, 0) + 1
                with open(out_filename, "w") as out_samfile:
                    out_sam = simplesam.Writer(out_samfile, in_sam.header)
                    out_sam.write(read)
            else:
                barcodes[cell_bc] += 1
                with open(out_filename, "a") as out_samfile:
                    print(str(read), end="", file=out_samfile)
    in_sam.close()

if __name__ == "__main__":
    main()


ddSeeker_barcodes.py
#!/usr/bin/env python3

import logging
import sys
import time
from argparse import ArgumentParser, ArgumentTypeError
import pysam
from pathlib import Path
from functools import partial
from multiprocessing import Pool
from itertools import islice
from Bio import pairwise2, SeqIO
from gzip import open as gzopen

logging.basicConfig(level=logging.INFO, datefmt='%H:%M:%S',
                    format="[%(asctime)s] %(levelname)s - %(message)s")
_match_table = [1, 0, -1, -1]
_linkers = ["TAGCCATCGCATTGC", "TACCTCTGAGCTGAA"]
_local_aligner = partial(pairwise2.align.localms, one_alignment_only=True)

class countDict(dict):
    """
    >>> x = countDict([("a",1), ("b",1)])
    >>> y = countDict([("a",1), ("b",2), ('c',10)])
    >>> x+y
    {'a': 2, 'c': 10, 'b': 3}

    >>> x = countDict([("a",1), ("b",2)])
    >>> print(x)
    {'a':1, 'b':2}
    >>> x.update(["a", "b", "c"])
    >>> print(x)
    {'a':2, 'c':1, 'b':3}

    """
    def update(self, iterable):
        for key in iterable:
            self[key] = self.get(key, 0) + 1

def extract_barcodes(read):
    starts = []
    for linker in _linkers:  # align the two linkers
        alignment = _local_aligner(read, linker, *_match_table)[0]
        check_length = alignment[4] - alignment[3] == len(linker)  # end - start
        check_match = alignment[2] >= len(linker) - 1  # alignment score (allow 1 mismatch)
        if (check_length and check_match):
            starts.append(alignment[3])
        else:
            starts.append("")

    if all(starts) and (starts[1]-starts[0] == 21) and starts[0] >= 6:
        bc_blocks = [str(read[starts[0] - 6: starts[0]]),
                     str(read[starts[1] - 6: starts[1]]),
                     str(read[starts[1] + 15: starts[1] + 21])]
        if any([len(_) != 6 for _ in bc_blocks]):
            bc_blocks = []
    else:
        bc_blocks = []

    return(bc_blocks)

def main(argv=None):
    if not argv:
        argv = sys.argv[1:]

    # Parameters ####
    args = parse_args(argv)
    in_files = args.input
    out_file = args.output
    cores = args.cores

    bc_count = countDict()

    # Processing ####
    _start = time.perf_counter()
    for in_file in in_files:
        ext = Path(in_file).suffixes
        if ext[-1] == ".bam" or ext[-1] == ".sam":
            file_type = "bam"
        elif ext[-2] == ".fastq" and ext[-1] == ".gz":
            file_type = "fastq"
        else:
            raise IOError("Unrecognized input file extension '{}'".format(ext[-1]))

        logging.info("Extracting barcodes from '{}'".format(in_file))
        if file_type == "bam":
            in_bam = pysam.AlignmentFile(in_file, check_sq=False)
            in_bam_iter = islice(in_bam.fetch(until_eof=True), None, None, 2)
            reads = (_.query_sequence.upper() for _ in in_bam_iter)
        else:
            reads = (str(_.seq.upper()) for _ in SeqIO.parse(gzopen(in_file, "rt"), format="fastq"))

        if args.subset: # DEBUGGING PURPOSES
            reads = islice(reads, args.subset)

        pool = Pool(cores)
        barcodes = pool.map(extract_barcodes, reads)
        for bc_blocks in barcodes:
            bc_count.update(bc_blocks)
        pool.close()

        if file_type == "bam":
            in_bam.close()

    sorted_bc = sorted(bc_count, key=lambda x: bc_count[x], reverse=True)
    with out_file.open("w") as out:
        for k in sorted_bc:
            out.write("{}\t{}\n".format(k, bc_count[k]))
    _end = time.perf_counter()
    logging.info("Done. Elapsed time: {} minutes.".format(round((_end-_start)/60, 2)))

def parse_args(args):
    description = "A tool to generate a set of barcode blocks " +\
            "to be used by ddSeeker.py"

    parser = ArgumentParser(description=description)

    parser.add_argument("-i", "--input", required=True, nargs="+",
            help = "Merged paired-end uBAM file(s) or fastq files")

    barcodes_file = Path(sys.argv[0]).resolve().parent.joinpath("barcodes.txt")
    parser.add_argument("-o", "--output", type=Path, default=barcodes_file,
            help = "File reporting barcodes from most represented to less represented" +\
                   "(default=/path-to-ddSeeker/barcode.txt)")

    parser.add_argument("-c","--cores", type=int, default=1,
            help="Number of processing units (CPUs) to use (default=1)")

    parser.add_argument("--subset", type=int,
        help="Select a lower number of reads to analyze [for debugging purposes]")

    parser.add_argument("-v", '--version', action='version', version='%(prog)s 0.9.0')
    args = parser.parse_args()

    return(args)

if __name__ == "__main__":
    main()


code/ddSeeker.py
#!/usr/bin/env python3

import logging
import pysam
import sys
import time
from argparse import ArgumentParser
from Bio import pairwise2
from Bio.SeqIO.QualityIO import FastqGeneralIterator
from functools import partial
from gzip import open as gzopen
from itertools import islice
from multiprocessing import Pool
from numpy import cumsum
from pathlib import Path
from re import match as re_match

logging.basicConfig(level=logging.INFO, datefmt='%H:%M:%S', format="[%(asctime)s] %(levelname)s - %(message)s")
_linkers = ["TAGCCATCGCATTGC", "TACCTCTGAGCTGAA"]
_barcodes = ["AAAGAA", "AACAGC", "AACGTG", "AAGCCA", "AAGTAT", "AATTGG",
        "ACAAGG", "ACCCAA", "ACCTTC", "ACGGAC", "ACTGCA", "AGACCC", "AGATGT",
        "AGCACG", "AGGTTA", "AGTAAA", "AGTCTG", "ATACTT", "ATAGCG", "ATATAC",
        "ATCCGG", "ATGAAG", "ATTAGT", "CAACCG", "CAAGTC", "CACCAC", "CACTGT",
        "CAGACT", "CAGGAG", "CATAGA", "CCACGC", "CCGATG", "CCGTAA", "CCTCTA",
        "CGAAAG", "CGAGCA", "CGCATA", "CGGCGT", "CGGTCC", "CGTTAT", "CTAGGT",
        "CTATTA", "CTCAAT", "CTGTGG", "CTTACG", "CTTGAA", "GAAATA", "GAAGGG",
        "GACTCG", "GAGCTT", "GAGGCC", "GAGTGA", "GATCAA", "GCCAGA", "GCCGTT",
        "GCGAAT", "GCGCGG", "GCTCCC", "GCTGAG", "GCTTGT", "GGACGA", "GGATTG",
        "GGCCAT", "GGGATC", "GGTAGG", "GGTGCT", "GTACAG", "GTCCTA", "GTCGGC",
        "GTGGTG", "GTTAAC", "GTTTCA", "TAAGCT", "TAATAG", "TACCGA", "TAGAGG",
        "TATTTC", "TCAGTG", "TCATCA", "TCCAAG", "TCGCCT", "TCGGGA", "TCTAGC",
        "TGAATT", "TGAGAC", "TGCGGT", "TGCTAA", "TGGCAG", "TGTGTA", "TGTTCG",
        "TTAAGA", "TTCGCA", "TTCTTG", "TTGCTC", "TTGGAT", "TTTGGG"]

_local_aligner = partial(pairwise2.align.localxs, one_alignment_only=True)
_global_aligner = partial(pairwise2.align.globalxs, score_only=True, one_alignment_only=True)

_cell_count={}
_error_count = {}

def hamming_dist(s1, s2):
    """Return the Hamming distance between equal-length sequences

    >>> hamming_distance("ACG", "ACT")
    1
    """
    if (len(s1) != len(s2)):
        raise ValueError("Undefined for sequences of unequal length")
    return sum(el1 != el2 for el1, el2 in zip(s1, s2))

def fix_block(block):
    """Set barcode to most similar (up to 1 mismatch)

    >>> fix_block("TTTGGG")
    'TTTGGG'

    >>> fix_block("TTTGAG")
    'TTTGGG'

    >>> fix_block("TTTGAA")
    ''
    """
    for bc in _barcodes:
        score = _global_aligner(bc, block, -1, -1)
        if len(block) >= 5 and score >= 5:
            return(bc)
    else:
        return(None)

def get_tags(record):
    """Extract barcodes from R1 and return a tuple of SAM-format TAGs.
    Default tag values:
    - XC = cell barcode
    - XM = molecular barcode (UMI)
    - XE = error

    Errors:
    LX = both linkers not aligned correctly
    L1 = linker 1 not aligned correctly
    L2 = linker 2 not aligned correctly
    I = indel in BC2
    D = deletion in Phase Block or BC1
    J = indel in BC3 or ACG trinucleotide
    K = indel in UMI or GAC trinucleotide
    B = one BC with more than 1 mismatch"""

    sequence = record[1].upper()
    quality = record[2]
    linker_start_index = []
    k = []
    # find the start of the 2 linker sequences, allowing 1 edit distance (mismatch, indel)
    for linker in _linkers:
        try:
            alignment = _local_aligner(sequence, linker, -2, -1)[0]
        except IndexError:
            linker_start_index.append(None)
            continue
        seqA, seqB, score, begin, end = alignment
        length = end - begin
        if length == 15 and score >= 14:
            # 0-1 mismatch
            linker_start_index.append(begin)
            k.append(0)
        elif length == 14 and score == 14:
            # 1 mismatch at starting position
            linker_start_index.append(begin - 1)
            k.append(0)
        elif "-" in seqA[begin:end] and length == 15 and score == 12:
            # 1 deletion
            linker_start_index.append(begin)
            k.append(-1)
        elif "-" in seqB and length == 16 and score == 13:
            # 1 insertion
            linker_start_index.append(begin)
            k.append(1)
        else:
            linker_start_index.append(None)

    if linker_start_index[0] is None and linker_start_index[1] is None:
        return(dict([(_tag_error, "LX")])) # no linker aligned
    elif linker_start_index[0] is None:
        return(dict([(_tag_error, "L1")])) # linker 1 not aligned
    elif linker_start_index[1] is None:
        return(dict([(_tag_error, "L2")])) # linker 2 not aligned

    # extract block 2 if the distance between the linker blocks is equal to 21 (+/- 1 indel)
    if linker_start_index[1]-linker_start_index[0] == 21+k[0]:
        bc2 = sequence[linker_start_index[1]-6: linker_start_index[1]]
        bc2_q = quality[linker_start_index[1]-6: linker_start_index[1]]
    elif linker_start_index[1]-linker_start_index[0] == 20+k[0]: # 1 deletion in bc2
        bc2 = sequence[linker_start_index[1]-5: linker_start_index[1]]
        bc2_q = sequence[linker_start_index[1]-5: linker_start_index[1]]
    elif linker_start_index[1]-linker_start_index[0] == 22+k[0]: # 1 insertion in bc2
        bc2 = sequence[linker_start_index[1]-7: linker_start_index[1]]
        bc2_q = quality[linker_start_index[1]-7: linker_start_index[1]]
    else:
        return(dict([(_tag_error, "I")]))

    # extract block 1 if the length of block 1 is greater than or equal to 5 bases
    if linker_start_index[0] < 5:
        return(dict([(_tag_error, "D")]))
    elif linker_start_index[0] == 5:
        bc1 = sequence[: linker_start_index[0]]
        bc1_q = quality[: linker_start_index[0]]
    else:
        bc1 = sequence[linker_start_index[0]-6: linker_start_index[0]]
        bc1_q = quality[linker_start_index[0]-6: linker_start_index[0]]

    # extract ACG block and check the sequence
    acg = sequence[linker_start_index[1]+21+k[1]: linker_start_index[1]+24+k[1]]
    i = 0
    thr = 0
    while i < len(acg) and thr <= 1:
        if acg[i] != "ACG"[i]:
            thr += 1
        i += 1
    else:
        if thr > 1:
            return(dict([(_tag_error, "J")]))
    try:
        dist_acg = hamming_dist(acg, "ACG")
    except ValueError:
        dist_acg = float("inf")
    if dist_acg > 1:
        return(dict([(_tag_error, "J")]))

    # extract GAC block and check the sequence
    gac = sequence[linker_start_index[1]+32+k[1]: linker_start_index[1]+35+k[1]]
    try:
        dist_gac = hamming_dist(gac, "GAC")
    except ValueError:
        dist_gac = float("inf")
    if dist_gac > 1:
        return(dict([(_tag_error, "K")]))

    # extract block 3
    bc3 = sequence[linker_start_index[1]+15+k[1]: linker_start_index[1]+21+k[1]]
    bc3_q = quality[linker_start_index[1]+15+k[1]: linker_start_index[1]+21+k[1]]

    # correct blocks 1,2,3 and compose cell barcode
    barcode = []
    for block in (bc1, bc2, bc3):
        fixed = fix_block(block)
        if fixed:
            barcode.append(fixed)
        else:
            return(dict([(_tag_error, "B")]))
    barcode = "".join(barcode)
    barcode_q = "".join([bc1_q, bc2_q, bc3_q])

    # extract Unique Molecula Identifier (UMI)
    umi = sequence[linker_start_index[1]+24+k[1]: linker_start_index[1]+32+k[1]]
    umi_q = quality[linker_start_index[1]+24+k[1]: linker_start_index[1]+32+k[1]]

    #  print(dict([(_tag_bc, barcode), (_tag_umi, umi), (_tag_umi_q, umi_q), (_tag_bc_q, barcode_q)]))
    return(dict([(_tag_bc, barcode), (_tag_umi, umi), (_tag_umi_q, umi_q), (_tag_bc_q, barcode_q)]))

def compute_summary(tags):
    # summary statistics
    if tags.get(_tag_error):
        _error_count[tags[_tag_error]] = _error_count.get(tags[_tag_error], 0) + 1
    elif tags.get(_tag_bc):
        _cell_count[tags[_tag_bc]] = _cell_count.get(tags[_tag_bc], 0) + 1
        _error_count["PASS"] = _error_count.get("PASS", 0) + 1

def write_summary(summary_path):
    file_name = summary_path + ".errors.csv"
    ordered_tags = ["LX", "L1", "L2", "I", "D", "J", "K", "B", "PASS"]
    out = open(file_name, "w")
    out.write("Error\tCount\tFraction\n")
    for tag in ordered_tags:
        count = _error_count.get(tag, 0)
        fraction = _error_count.get(tag, 0)/sum(_error_count.values())
        out.write("{}\t{}\t{}\n".format(tag, count, fraction))
    out.close()

    file_name = summary_path + ".cell_barcodes.csv"
    sorted_barcodes = sorted(_cell_count, key=lambda x: _cell_count[x], reverse=True)
    cell_cumsum = cumsum([_cell_count[b] for b in sorted_barcodes])/sum(_cell_count.values())
    out = open(file_name, "w")
    out.write("Cell_Barcode\tCount\tCumulative_Sum\n")
    for i, barcode in enumerate(sorted_barcodes):
        out.write("{}\t{}\t{}\n".format(barcode, _cell_count[barcode],
            cell_cumsum[i]))
    out.close()

def main():
    args = parse_args()

    global _tag_bc
    global _tag_bc_q
    global _tag_umi
    global _tag_umi_q
    global _tag_error
    _tag_bc    = args.tag_bc
    _tag_bc_q  = args.tag_bc_q
    _tag_umi   = args.tag_umi
    _tag_umi_q = args.tag_umi_q
    _tag_error = args.tag_error

    bam_write_mode = "w" if args.output == "-" else "wb"
    in_filename1, in_filename2 = args.input

    # Processing ####
    _start = time.perf_counter()
    in_reads1 = (record for record in FastqGeneralIterator(gzopen(in_filename1, "rt")))
    in_reads2 = (record for record in FastqGeneralIterator(gzopen(in_filename2, "rt")))

    if args.subset: # DEBUGGING PURPOSES
        in_reads1 = islice(in_reads1, args.subset)

    if args.pipeline == "dropseq":
        bam_header = {'HD':{'VN': '1.6', 'SO':'unknown'}}
        out_bam = pysam.AlignmentFile(args.output, bam_write_mode, header=bam_header)

    elif args.pipeline == "scpipe":
        out_fastq = gzopen(args.output, "wt")

    logging.info("Extracting tags.")
    pool = Pool(args.cores)

    for (i, tags) in enumerate(pool.imap(get_tags, in_reads1), 1):
        if (i) % 1e6 == 0:
            logging.info("{} reads processed.".format(str(i)))
        if args.pipeline == "dropseq":
            title, seq, qual = next(in_reads2)
            sam_record = pysam.AlignedSegment()
            sam_record.query_name = title.split()[0]
            sam_record.query_sequence = seq
            sam_record.query_qualities = pysam.qualitystring_to_array(qual)
            sam_record.template_length = len(seq)
            sam_record.flag = 4
            sam_record.set_tags(tags.items())
            out_bam.write(sam_record)
        elif args.pipeline == "scpipe" and tags.get(_tag_bc):
            title, seq, qual = next(in_reads2)
            out_fastq.write("@{}_{}#{}\n{}\n+\n{}\n".format(tags[_tag_bc], tags[_tag_umi], title.split()[0], seq, qual))

        if args.summary_prefix: # summary statistics
            compute_summary(tags)
    logging.info("{} reads processed.".format(str(i)))
    pool.close()
    logging.info("All reads analyzed.")

    if args.pipeline == "dropseq":
        out_bam.close()
    elif args.pipeline == "scpipe":
        out_fastq.close()

    if args.summary_prefix:
        logging.info("Writing summary files.")
        write_summary(args.summary_prefix)

    _end = time.perf_counter()
    logging.info("Done. Elapsed time: {} minutes.".format(round((_end-_start)/60, 2)))

def parse_args():
    description = "A tool to extract cellular and molecular identifiers " +\
        "from single cell RNA sequencing experiments"

    parser = ArgumentParser(description=description)

    parser.add_argument("-i", "--input", nargs=2,
            help="R1 and R2 from a paired end sequencing experiment")
    parser.add_argument("-o", "--output", required=True,
            help="Tagged unmapped BAM file (use '-' to output to stdout)")

    parser.add_argument("--pipeline", default="dropseq", choices=["dropseq", "scpipe"],
        help="Set output type depending on pipeline tool chosen")

    parser.add_argument("-s", "--summary-prefix",
        help="Prefix for summary files (including absolute or relative paths)")

    parser.add_argument("-c","--cores", type=int, default=1,
        help="Number of processing units (CPUs) to use (default=1)")

    parser.add_argument("--tag-bc", type=str, default="XC",
        help="Tag for single cell barcode (default=XC)")

    parser.add_argument("--tag-bc-q", type=str, default="XQ",
        help="Tag for single cell barcode base quality (default=XQ)")

    parser.add_argument("--tag-umi", type=str, default="XM",
        help="Tag for Unique Molecular Identifier (default=XM)")

    parser.add_argument("--tag-umi-q", type=str, default="Xq",
        help="Tag for Unique Molecular Identifier base quality (default=Xq)")

    parser.add_argument("--tag-error", type=str, default="XE",
        help="Tag for errors (default=XE)")

    parser.add_argument("--subset", type=int,
        help="Select a lower number of reads to analyze [debugging]")

    parser.add_argument("-v", '--version', action='version', version='%(prog)s 1.2.1')

    args = parser.parse_args()

    # check parameters
    if not (args.tag_bc != args.tag_umi != args.tag_error != args.tag_bc):
        logging.error("Tags provided with '--tag' must be all different.")
        sys.exit(1)

    tag_pattern = "^[A-Za-z][A-Za-z0-90]$"
    if not (re_match(tag_pattern, args.tag_bc) and re_match(tag_pattern, args.tag_umi) and re_match(tag_pattern, args.tag_error)):
        logging.error("Tags provided with '--tag' must be two-character strings matching /[A-Za-z][A-Za-z0-9]/")
        sys.exit(1)

    for x in args.input:
        ext = "".join(Path(x).suffixes)
        if ext != ".fastq.gz":
            logging.error("Invalid input file extensions '{}': '.fastq.gz' is required".format(ext))
            sys.exit(1)

    return(args)

if __name__ == "__main__":
    main()
